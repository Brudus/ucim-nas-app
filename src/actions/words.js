import database from '../firebase/firebase';

// ADD_WORD
export const addWord = (word) => ({
    type: 'ADD_WORD',
    word
});

export const startAddWord = (wordData = {}) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        const {
            source = '', 
            destination= '', 
            repeatAt = 0,
            interval = 0,
            reps = 0,
            easeFactor = 2.5,
            repeatAtInverted = 0,
            intervalInverted = 0,
            repsInverted = 0,
            easeFactorInverted = 2.5
        } = wordData;
        const word = { source, destination, repeatAt, interval, reps, easeFactor, repeatAtInverted, intervalInverted, repsInverted, easeFactorInverted };
        
        return database.ref(`users/${uid}/words`).push(word).then((ref) => {
            dispatch(addWord({
                id: ref.key,
                ...word
            }));
        });
    };
};

// REMOVE_WORD
export const removeWord = ({ id } = {}) => ({
    type: 'REMOVE_WORD',
    id
});

export const startRemoveWord = ({ id } = {}) => {
    return (dispatch,getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/words/${id}`).remove().then(() => {
            dispatch(removeWord({ id }));
        });
    };
};

// EDIT_WORD
export const editWord = (id, updates) => ({
    type: 'EDIT_WORD',
    id,
    updates
});

export const startEditWord = (id, updates) => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/words/${id}`).update(updates).then(() => {
            dispatch(editWord(id, updates));
        })
    };
};

// SET_WORDS
export const setWords = (words) => ({
    type: 'SET_WORDS',
    words
});

export const startSetWords = () => {
    return (dispatch, getState) => {
        const uid = getState().auth.uid;
        return database.ref(`users/${uid}/words`).once('value').then((snapshot) => {
            const words = [];

            snapshot.forEach((childSnapshot) => {
                words.push({
                    id: childSnapshot.key,
                    ...childSnapshot.val()
                });
            });
            
            dispatch(setWords(words));
        });
    };
};