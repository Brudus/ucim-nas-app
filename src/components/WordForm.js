import React from 'react';

export default class WordForm extends React.Component {
    constructor(props) {
        super(props);
        this.state = {
            destination: props.word ? props.word.destination : '',
            source: props.word ? props.word.source : '',
            repeatAt: props.word ? props.word.repeatAt : undefined,
            repeatAtInverted: props.word ? props.word.repeatAtInverted : undefined,
            error: ''
        };
    }

    onDestinationChange = (e) => {
        const destination = e.target.value;
        this.setState(() => ({ destination }));
    };

    onSourceChange = (e) => {
        const source = e.target.value;
        this.setState(() => ({ source }));
    }

    onSubmit = (e) => {
        e.preventDefault();

        if (!this.state.destination || !this.state.source) {
            const error = 'Please provide a word and its translation in your language.'
            this.setState(() => ({ error }));
        } else {
            const error = '';
            this.setState(() => ({ error }));
            this.props.onSubmit({
                destination: this.state.destination,
                source: this.state.source,
                repeatAt: this.state.repeatAt ? this.state.repeatAt : Date.now(),
                repeatAtInverted: this.state.repeatAtInverted ? this.state.repeatAtInverted : Date.now()
            });
            this.setState(() => ({
                destination: '',
                source: '',
                repeatAt: undefined,
                repeatAtInverted: undefined
            }));
        }
    };

    render() {
        return (
            <div>
                {this.state.error && <p>{this.state.error}</p>}
                <form onSubmit={this.onSubmit}>
                    <input
                        type="text"
                        placeholder="Word in Serbo-Croatian"
                        autoFocus
                        value={this.state.destination}
                        onChange={this.onDestinationChange}
                    />
                    <input
                        type="text"
                        placeholder="Word in German"
                        value={this.state.source}
                        onChange={this.onSourceChange}
                    ></input>
                    <button>{ this.props.word ? 'Save' : 'Add' }</button>
                </form>
            </div>
        );
    };
}