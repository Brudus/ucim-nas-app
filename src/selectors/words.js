export default (words, { text, sortBy }) => {
    return words.filter((word) => {
        const sourceText = word.source.toLowerCase()
        const destinationText = word.destination.toLowerCase();
        const searchText = text.toLowerCase().trim();

        const sourceMatch = sourceText.includes(searchText);
        const destinationMatch = destinationText.includes(searchText);
        return sourceMatch || destinationMatch;
    }).sort((a, b) => {
        if (sortBy === 'repeatAt') {
            return a.repeatAt > b.repeatAt ? 1 : -1;
        } else if (sortBy === 'source') {
            const aText = a.source.toLowerCase().trim();
            const bText = b.source.toLowerCase().trim();
            return aText > bText ? 1 : -1;
        } else if (sortBy === 'destination') {
            const aText = a.destination.toLowerCase().trim();
            const bText = b.destination.toLowerCase().trim();
            return aText > bText ? 1 : -1;
        }
    });
};