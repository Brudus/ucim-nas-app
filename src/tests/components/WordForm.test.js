import React from 'react';
import { shallow } from 'enzyme';
import WordForm from '../../components/WordForm';
import words from '../fixtures/words';

test('should render WordForm correctly', () => {
    const wrapper = shallow(<WordForm />);
    expect(wrapper).toMatchSnapshot();
});

test('should render WordForm with word data', () => {
    const wrapper = shallow(<WordForm word={words[0]}/>);
    expect(wrapper).toMatchSnapshot();
});

test('should render error for invalid form submission', () => {
    const wrapper = shallow(<WordForm />);
    expect(wrapper).toMatchSnapshot();
    wrapper.find('form').simulate('submit', { 
        preventDefault: () => {} 
    });
    expect(wrapper.state('error').length).toBeGreaterThan(0);
    expect(wrapper).toMatchSnapshot();
});

test('should set destination on input change', () => {
    const value = 'New destination';
    const wrapper = shallow(<WordForm />);
    wrapper.find('input').at(0).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('destination')).toBe(value);
});

test('should set source on input change', () => {
    const value = 'New source';
    const wrapper = shallow(<WordForm />);
    wrapper.find('input').at(1).simulate('change', {
        target: { value }
    });
    expect(wrapper.state('source')).toBe(value);
});

test('should call onSubmit prop for valid from submission', () => {
    const onSubmitSpy = jest.fn();
    const wrapper = shallow(<WordForm word={words[1]} onSubmit={onSubmitSpy} />);
    wrapper.find('form').simulate('submit', { 
        preventDefault: () => {} 
    });
    expect(wrapper.state('error')).toBe('');
    expect(onSubmitSpy).toHaveBeenLastCalledWith({
        destination: words[1].destination,
        source: words[1].source,
        repeatAt: words[1].repeatAt
    });
});