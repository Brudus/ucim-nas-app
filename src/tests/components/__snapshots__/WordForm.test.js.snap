// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render WordForm correctly 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Word in Serbo-Croatian"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Word in German"
      type="text"
      value=""
    />
    <button>
      Add
    </button>
  </form>
</div>
`;

exports[`should render WordForm with word data 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Word in Serbo-Croatian"
      type="text"
      value="dobro"
    />
    <input
      onChange={[Function]}
      placeholder="Word in German"
      type="text"
      value="gut"
    />
    <button>
      Save
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 1`] = `
<div>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Word in Serbo-Croatian"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Word in German"
      type="text"
      value=""
    />
    <button>
      Add
    </button>
  </form>
</div>
`;

exports[`should render error for invalid form submission 2`] = `
<div>
  <p>
    Please provide a word and its translation in your language.
  </p>
  <form
    onSubmit={[Function]}
  >
    <input
      autoFocus={true}
      onChange={[Function]}
      placeholder="Word in Serbo-Croatian"
      type="text"
      value=""
    />
    <input
      onChange={[Function]}
      placeholder="Word in German"
      type="text"
      value=""
    />
    <button>
      Add
    </button>
  </form>
</div>
`;
